File: src/parsers/pdf_parser.py
"""
PDF parser: extract text using pypdf.
"""

from pathlib import Path

from pypdf import PdfReader


def parse_pdf(file_path: str) -> str:
    """
    Extract text from PDF at file_path. Returns concatenated text from all pages.
    """
    path = Path(file_path)
    if not path.exists() or not path.is_file():
        raise FileNotFoundError(f"PDF not found: {file_path}")
    reader = PdfReader(str(path))
    parts = []
    for page in reader.pages:
        text = page.extract_text()
        if text:
            parts.append(text)
    return "\n".join(parts)

File: src/parsers/pdf_parser.cpp
#include "document_ingestion/parsers_internal.hpp"
#include <filesystem>
#include <stdexcept>

#ifdef ENABLE_POPPLER
#include <poppler-document.h>
#include <poppler-page.h>
#endif

namespace document_ingestion {

std::string parse_pdf(const std::string& file_path) {
  std::filesystem::path p(file_path);
  if (!std::filesystem::exists(p) || !std::filesystem::is_regular_file(p))
    throw std::runtime_error("PDF not found: " + file_path);

#ifdef ENABLE_POPPLER
  auto doc = poppler::document::load_from_file(file_path);
  if (!doc) throw std::runtime_error("Failed to load PDF: " + file_path);
  std::string result;
  for (int i = 0; i < doc->pages(); ++i) {
    auto page = doc->create_page(i);
    if (page) {
      result += page->text().to_latin1();
      if (i + 1 < doc->pages()) result += "\n";
    }
  }
  return result;
#else
  (void)file_path;
  throw std::runtime_error("PDF parsing requires build with -DENABLE_PDF=ON and Poppler");
#endif
}

}  // namespace document_ingestion
